Fix "ESC to hide" functionality

bracketttc opened the pull request:
Main goal was to fix the hiding functionality. Other things bothered me along the way.

Changes boil down to:
* `/etc/classification-banner/banner.conf` file is no longer required. Previously, the script would error out in a manner not immediately apparent that it was not finding a configuration file (or was given an improper configuration file).
* Fixed fallback to GTK for screen size detection. `xrandr` is still attempted to be called three times if it's not present or fails for any reason though.
* Added some warnring prints for improper config/arguments.
* Fixed ESC to hide functionality
* Fixed opacity setting

Tested on Fedora 38.

Closes #12 
Closes #14 
Closes #16

bracketttc committed:
Allow banner to open without a config file

bracketttc committed:
Remove unnecessary debug print

bracketttc committed:
Fix fallback to GTK screen detection

bracketttc committed:
Warn users of unrecognized config options

bracketttc committed:
Warn user if hres XOR vres specified

bracketttc committed:
Fix event callback signatures

bracketttc committed:
Rework 'ESC to hide' functionality

Uses Python's threading module to wait off of the UI thread to restore
the hidden banner. This change also introduces a boolean to the
ClassificationBanner class that tracks whether it was intentionally
hidden.

bracketttc committed:
Fix monitors-changed callback registration

bracketttc commented:
Tested on RHEL7 machine with slower cores and something's causing it to use way too much CPU after restoring from being hidden.

bracketttc committed:
Fix CPU usage issue caused by use of `idle_add`

Improperly introduced a use of `idle_add` instead of `idle_add_once`
(not available in GObject apparently) which caused a banner to
continuously run `restore` after being intentionally hidden by hitting
ESC. My understanding is that calling a GUI method from a non-GUI thread
may cause unpredictable results in some cases, but without
`idle_add_once` the alternatives are more involved.

Backing up further, use of 'threading` entirely could be replaced by
using `timeout_add_once` if that was supported by GObject.

bracketttc committed:
Correctly add time-delayed restore

PyGObject lacking `idle_add_once` and `timeout_add_once`, the
appropriate pattern is to use `timeout_add` and return `False` from the
called function which signals the main loop to remove the timeout
instead of running it again when the timeout expires.

Note that the units for `timeout_add`'s interval is milliseconds where
both `time.sleep` and `threading.Timer` are both in seconds.

bracketttc commented:
Tested on Fedora 38 and RHEL7.

CPU usage is back to idle/background levels because I'm not accidentally making it a busy loop.

bracketttc committed:
Fix linter diagnostics

bracketttc committed:
Fix typo setting window opacity

bracketttc commented:
@redhatrises can you take a look at this?

bracketttc committed:
Make unrecognized options a real f-string

redhatrises commented:
I think I am fine with many of these changes. There are things to do like move to log instead of print, don't allow the user to run a banner over an existing one, seeing if the banner can be on top of the menu/bottom bar.... but that's outside the scope of this PR. One thing I am wondering (without having looked at the gtk docs) is if the text can stay the same opacity while the banner's opacity changes as the text can be hard to read in some cases. I get the feeling that it might not be the case.

redhatrises commented:
Also seems as if the banner disappears behind windows instead of being overlaid on top of the windows which is less than ideal (unless configured to do so.)

bracketttc commented:
I ran into that, but I thought I had it sorted out. I'll have to take another look.

bracketttc commented:
> I think I am fine with many of these changes. There are things to do like move to log instead of print, don't allow the user to run a banner over an existing one, seeing if the banner can be on top of the menu/bottom bar.... but that's outside the scope of this PR. One thing I am wondering (without having looked at the gtk docs) is if the text can stay the same opacity while the banner's opacity changes as the text can be hard to read in some cases. I get the feeling that it might not be the case.

At some level, this is a "best-effort" for the general case of the requirement. If you've got better defined criteria, you may be able to do better. For instance, if you're only using `gdm` as your desktop manager, and `gnome-shell` as your only available window manager in your deployment, you should almost certainly prefer using the gnome-shell extension as it's more tightly bound with those pieces and therefore better able to work with and around them. Red Hat's official documentation for [RHEL8](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/using_the_desktop_environment_in_rhel_8/assembly_displaying-the-system-security-classification_using-the-desktop-environment-in-rhel-8) and [RHEL9](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/customizing_the_gnome_desktop_environment/assembly_displaying-the-system-security-classification_customizing-the-gnome-desktop-environment) actually points readers to that solution.

> I ran into that, but I thought I had it sorted out. I'll have to take another look.

I can't repro this issue anymore and adding a hook to track changes to `Gdk.WindowState` doesn't seem to show the `GDK_WINDOW_STATE_ABOVE` flag turning off. But my personal machine isn't a representative case because it's Fedora 38. I can poke it during the week on some of work's RHEL7 and RHEL8 machines.

bracketttc committed:
Force stay-on-top back on if lost

bracketttc committed:
Remove unnecessary try-excepts

`Gtk.Window.connect` is called immediately prior and if we're trying to
protect against a version of GTK that doesn't include the signal we're
hooking onto, that's not the exception thrown.

bracketttc commented:
Ah, merge conflict. Is this a merge-out or rebase kind of repo?

bracketttc commented:
Rebased.
